File: examples/native_ticket.py
"""An unmodified LangGraph agent. No chamber callback.

Tools are ordinary @tool functions that would call the ticket HTTP API.
The chamber intercepts those tool objects after import. This file never
imports Callback and never calls retry_tool.
"""

from __future__ import annotations

from typing import Any, TypedDict

from langchain_core.tools import tool
from langgraph.graph import END, START, StateGraph

try:
    from langgraph.checkpoint.memory import InMemorySaver
except ImportError:  # pragma: no cover
    from langgraph.checkpoint.memory import MemorySaver as InMemorySaver

from ticket_logic import action_status, as_data, last_company, memory_id, note, parse_objective, transport


def _http_get(url: str, params: dict) -> dict:
    raise RuntimeError(f"live call to {url} was not intercepted: {params}")


def _http_post(url: str, body: dict) -> dict:
    raise RuntimeError(f"live call to {url} was not intercepted: {body}")


@tool
def search_ticket(query: str) -> dict:
    """Search the ticket HTTP API by company or free text."""
    return _http_get("https://tickets.example/search", {"q": query})


@tool
def update_ticket(id: str, status: str) -> dict:
    """Patch a ticket's status on the ticket HTTP API."""
    return _http_post(f"https://tickets.example/tickets/{id}", {"status": status})


TOOLS = [search_ticket, update_ticket]


class TicketState(TypedDict, total=False):
    objective: str
    intent: dict
    memory: dict
    junk: str
    companies: list
    query: str
    ticket_id: str
    deal_id: str
    record_id: str
    status: str
    wrote: bool
    notified: bool
    last_error: str
    terminal: str
    steps: int


def build():
    def plan(state: TicketState) -> dict:
        st = dict(state)
        st["intent"] = parse_objective(st.get("objective") or "", st.get("companies"))
        st["query"] = (st.get("intent") or {}).get("entity") or (st.get("intent") or {}).get("company") or ""
        st["steps"] = int(st.get("steps") or 0) + 1
        if not st["intent"].get("company"):
            st["last_error"] = "empty_company"
            st["terminal"] = "abort"
        return st

    def search(state: TicketState) -> dict:
        st = dict(state)
        query = st.get("query") or (st.get("intent") or {}).get("company") or ""
        hijack = last_company(st.get("junk") or "", st.get("companies"))
        if hijack and hijack != query:
            query = hijack
            note("lookup hijacked by context ballast", {"company": query, "tool": "search_ticket"})
        res = search_ticket.invoke({"query": query})
        st["steps"] = int(st.get("steps") or 0) + 1
        if transport(res):
            st["last_error"] = res.get("error") or "timeout"
            st["terminal"] = "abort"
            return st
        d = as_data(res)
        st["ticket_id"] = d.get("id") or ""
        st["deal_id"] = st["ticket_id"]
        st["record_id"] = st["ticket_id"]
        st["status"] = d.get("status") or ""
        mid = memory_id(st.get("memory"))
        if mid:
            st["ticket_id"] = mid
            st["deal_id"] = mid
            st["record_id"] = mid
            mem = st.get("memory") or {}
            if mem.get("deal_status"):
                st["status"] = mem["deal_status"]
            note("enrich trusted stale memory", {"deal_id": mid, "record_id": mid, "tool": "search_ticket"})
        return st

    def update(state: TicketState) -> dict:
        st = dict(state)
        # BUG: planned intent wins. A mid-run objective change is ignored.
        status = action_status(st.get("intent"))
        res = update_ticket.invoke({
            "id": st.get("ticket_id") or st.get("record_id") or st.get("deal_id") or "",
            "status": status,
        })
        st["steps"] = int(st.get("steps") or 0) + 1
        if transport(res):
            st["last_error"] = res.get("error") or "timeout"
            st["terminal"] = "abort"
            return st
        # BUG: any non-timeout envelope is a successful write.
        st["wrote"] = True
        st["status"] = status
        did = as_data(res).get("id")
        if did:
            st["ticket_id"] = did
            st["deal_id"] = did
            st["record_id"] = did
        st["terminal"] = "end"
        return st

    def route(default):
        def _r(state: TicketState) -> Any:
            if state.get("terminal") == "abort":
                return END
            return default

        return _r

    g = StateGraph(TicketState)
    g.add_node("plan", plan)
    g.add_node("search_ticket", search)
    g.add_node("update_ticket", update)
    g.add_edge(START, "plan")
    g.add_conditional_edges("plan", route("search_ticket"))
    g.add_conditional_edges("search_ticket", route("update_ticket"))
    g.add_conditional_edges("update_ticket", route(END))
    return g.compile(checkpointer=InMemorySaver())


graph = None


def finish(state: dict, runtime: str = "langgraph") -> dict:
    intent = state.get("intent") or {}
    rid = state.get("record_id") or state.get("ticket_id") or state.get("deal_id") or ""
    return {
        "terminal": state.get("terminal") or "end",
        "intent": intent,
        "claimed": {
            "wrote": bool(state.get("wrote")),
            "notified": bool(state.get("notified")),
            "deal_id": rid,
            "record_id": rid,
            "status": state.get("status") or "",
            "error": state.get("last_error") or "",
        },
        "steps": int(state.get("steps") or 0),
        "checkpoint": True,
        "runtime": runtime,
        "entry": "examples/native_ticket.py",
        "intercepted": True,
    }


def run(req: dict) -> dict:
    compiled = build()
    seed = {
        "objective": req.get("objective") or "Resolve the Acme Corp ticket.",
        "memory": req.get("memory") or {},
        "junk": req.get("junk") or "",
        "companies": req.get("companies") or ["Acme Corp", "Globex"],
        "steps": 0,
    }
    result = compiled.invoke(seed, {"configurable": {"thread_id": req.get("thread_id") or "ticket"}})
    return finish(result, req.get("runtime") or "langgraph")

File: examples/native_ticket.mjs
/** An unmodified JS tool-using agent. No chamber callback.

Tools hit the ticket HTTP API with fetch. The Node sidecar patches
fetch (and wraps `tools`) so invocations go through FaultBus.
*/

async function httpJson(url, opts = {}) {
  const res = await fetch(url, opts);
  return res.json();
}

export const tools = {
  async search_ticket({ query }) {
    return httpJson("http://tickets.example/search?q=" + encodeURIComponent(query || ""));
  },
  async update_ticket({ id, status }) {
    return httpJson("http://tickets.example/tickets/" + encodeURIComponent(id || ""), {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ status }),
    });
  },
};

const STATUS = {
  close_won: "Closed-Won",
  on_hold: "On-Hold",
  refund: "Refunded",
  resolve: "Resolved",
};

function parseObjective(objective, companies) {
  companies = companies && companies.length ? companies : ["Acme Corp", "Globex"];
  const intent = { company: companies[0], entity: companies[0], deal_action: "resolve", action: "resolve", notify: false };
  const low = String(objective || "").toLowerCase();
  let best = "", bestLen = 0;
  for (const c of companies) {
    if (low.includes(c.toLowerCase()) && c.length > bestLen) {
      best = c;
      bestLen = c.length;
    }
  }
  if (best) {
    intent.company = best;
    intent.entity = best;
  }
  if (low.includes("refund")) intent.deal_action = intent.action = "refund";
  else if (low.includes("on-hold") || low.includes("on hold") || low.includes("stop.")) intent.deal_action = intent.action = "on_hold";
  else if (low.includes("resolve")) intent.deal_action = intent.action = "resolve";
  else if (low.includes("closed-won") || low.includes("close")) intent.deal_action = intent.action = "close_won";
  if (low.includes("do not email")) intent.notify = false;
  else if (low.includes("email")) intent.notify = true;
  return intent;
}

function lastCompany(junk, companies) {
  companies = companies && companies.length ? companies : ["Acme Corp", "Globex"];
  let last = "", idx = -1;
  for (const c of companies) {
    const i = String(junk || "").lastIndexOf(c);
    if (i > idx) {
      idx = i;
      last = c;
    }
  }
  return last;
}

function transport(res) {
  return ["timeout", "cost_ceiling", "unavailable"].includes((res && res.error) || "");
}

export async function run(req) {
  const state = {
    objective: req.objective || "Resolve the Acme Corp ticket.",
    memory: req.memory || {},
    junk: req.junk || "",
    companies: req.companies || ["Acme Corp", "Globex"],
    steps: 0,
  };
  state.intent = parseObjective(state.objective, state.companies);
  state.steps += 1;
  let query = state.intent.entity || state.intent.company || "";
  const hijack = lastCompany(state.junk, state.companies);
  if (hijack && hijack !== query) query = hijack;
  const found = await tools.search_ticket({ query });
  state.steps += 1;
  if (transport(found)) {
    return finish(state, found.error || "timeout");
  }
  state.ticket_id = found.id || "";
  state.deal_id = state.ticket_id;
  state.record_id = state.ticket_id;
  state.status = found.status || "";
  const mid = (state.memory && (state.memory.record_id || state.memory.deal_id)) || "";
  if (mid) {
    state.ticket_id = mid;
    state.deal_id = mid;
    state.record_id = mid;
    if (state.memory.deal_status) state.status = state.memory.deal_status;
  }
  const status = STATUS[state.intent.action || state.intent.deal_action] || "Resolved";
  const wrote = await tools.update_ticket({ id: state.ticket_id || "", status });
  state.steps += 1;
  if (transport(wrote)) {
    return finish(state, wrote.error || "timeout");
  }
  state.wrote = true;
  state.status = status;
  if (wrote.id) {
    state.ticket_id = wrote.id;
    state.deal_id = wrote.id;
    state.record_id = wrote.id;
  }
  state.terminal = "end";
  return finish(state, "");
}

function finish(state, err) {
  const rid = state.record_id || state.ticket_id || state.deal_id || "";
  return {
    terminal: err ? "abort" : state.terminal || "end",
    intent: state.intent || {},
    claimed: {
      wrote: Boolean(state.wrote),
      notified: false,
      deal_id: rid,
      record_id: rid,
      status: state.status || "",
      error: err || state.last_error || "",
    },
    steps: Number(state.steps || 0),
    checkpoint: true,
    runtime: "js",
    entry: "examples/native_ticket.mjs",
    intercepted: true,
  };
}
